WIP: Implement Filter POSTing and GETting

jcgruenhage opened the pull request:
This depends on matrix-org/gomatrix#42.

jcgruenhage commented on src/github.com/matrix-org/dendrite/clientapi/auth/storage/accounts/filter_table.go:
```diff
@@ -0,0 +1,92 @@
+	-- The filter
+	filter TEXT NOT NULL,
+	-- The ID
+	id SERIAL UNIQUE,
```
SERIAL means that postgres will just increment that forever, which leaks a tiny bit of data to users creating filters, namely how many filters have been created on the server. Synapse does something similar, but per user.

NegativeMjark replied:
Seems reasonable. (If we wanted to be clever here we could use a SHA-256 hash of the filter as the filter ID. That way we wouldn't get duplicates if the clients uploaded the same filter multiple times.)

jcgruenhage replied:
We would still get duplicates for multiple users though.. That this is at all user specific is somewhat displeasing, it would be nicer if there was just an ID and the associated filter, completely leaving out the users. @NegativeMjark Do you think I should use SHA256 as the ID or keep it as it is currently?

erikjohnston replied:
I think ID is fine for now

jcgruenhage commented on src/github.com/matrix-org/dendrite/clientapi/auth/storage/accounts/storage.go:
```diff
@@ -314,3 +319,21 @@ func (d *Database) GetThreePIDsForLocalpart(
+	if err != nil {
+		return "", err
+	}
+	return d.filter.findMaxID(ctx, localpart)
```
This is somewhat ugly, but I haven't found another way to get the integer created by postgres.

NegativeMjark replied:
You can use ```RETURNING``` to get the ID chosen by the ```INSERT``` https://www.postgresql.org/docs/9.5/static/dml-returning.html
There's an example of use doing this in dendrite at https://github.com/matrix-org/dendrite/blob/master/src/github.com/matrix-org/dendrite/syncapi/storage/output_room_events_table.go#L56

jcgruenhage replied:
I'll definitely do this, if we stick to using "SERIAL" here. If we use the hash as the ID, this is would not be needed anymore.

jcgruenhage commented on src/github.com/matrix-org/dendrite/clientapi/routing/routing.go:
```diff
@@ -185,23 +185,17 @@ func Setup(
+			vars := mux.Vars(req)
+			return writers.PutFilter(req, device, accountDB, vars["userId"])
 		}),
-	).Methods("POST", "OPTIONS")
```
What was this needed for? I haven't found that in the spec.

NegativeMjark replied:
We need the OPTIONS to satisfy CORS request pre-flighting https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request otherwise it's hard to hit the API from web clients.

jcgruenhage replied:
oh, okay, so I'll readd this. Should I go ahead and edit the current commit, or are multiple commits maybe better? The first would make the git history nicer, the latter would make the review and evolution inside this PR look nicer.

erikjohnston replied:
We squash merge so it doesn't really matter :)

jcgruenhage replied:
okay, nice 👍

jcgruenhage replied:
About the OPTIONS thingy there, I do not need to add more than this additional method, do I? I couldn't find more handling of OPTIONS at some other points, so I guess the go http package just handles that on his own?

erikjohnston replied:
No, you don't need to do anything more. It gets handled by the wrapper i think

erikjohnston commented:
The travis test failure is now due to linting errors:

```
src/github.com/matrix-org/dendrite/clientapi/writers/filter.go:30:1:warning: filterResponse is unused (deadcode)
src/github.com/matrix-org/dendrite/clientapi/auth/storage/accounts/storage.go:331:1:warning: exported method Database.PutFilter should have comment or be unexported (golint)
src/github.com/matrix-org/dendrite/clientapi/readers/filter.go:30:1:warning: exported function GetFilter should have comment or be unexported (golint)
src/github.com/matrix-org/dendrite/clientapi/writers/filter.go:34:1:warning: exported function PutFilter should have comment or be unexported (golint)
src/github.com/matrix-org/dendrite/clientapi/readers/filter.go:33:19:warning: 2 other occurrence(s) of "GET" found in: src/github.com/matrix-org/dendrite/clientapi/readers/login.go:63:19 src/github.com/matrix-org/dendrite/clientapi/readers/profile.go:52:19 (goconst)
src/github.com/matrix-org/dendrite/clientapi/readers/login.go:63:19:warning: 2 other occurrence(s) of "GET" found in: src/github.com/matrix-org/dendrite/clientapi/readers/filter.go:33:19 src/github.com/matrix-org/dendrite/clientapi/readers/profile.go:52:19 (goconst)
src/github.com/matrix-org/dendrite/clientapi/readers/profile.go:52:19:warning: 2 other occurrence(s) of "GET" found in: src/github.com/matrix-org/dendrite/clientapi/readers/filter.go:33:19 src/github.com/matrix-org/dendrite/clientapi/readers/login.go:63:19 (goconst)
src/github.com/matrix-org/dendrite/clientapi/writers/filter.go:30:6:warning: type filterResponse is unused (U1000) (megacheck)
src/github.com/matrix-org/dendrite/clientapi/writers/filter.go:62:2:warning: empty branch (SA9003) (megacheck)
src/github.com/matrix-org/dendrite/clientapi/writers/filter.go:67:2:warning: empty branch (SA9003) (megacheck)
```

jcgruenhage commented:
I'll fix those after fixing the missing TODOs above.

erikjohnston commented:
I think so long as you handle the two errors of the form:

```
	if err != nil {
		//TODO
	}
```

I'm happy to merge as is (and adding issues to track the TODOs)

jcgruenhage commented:
I've already fixed that locally (at home, I'm at work currently), and I've fixed everything the linter complains about. I'll rebase it once I'm home and then push the changes here, so that you can merge them.

jcgruenhage commented:
ehm....
what happened?!

jcgruenhage commented:
Apparently github autocloses PRs, when you acidentally force push the destination branch to the source branch...

erikjohnston commented:
> Apparently github autocloses PRs, when you acidentally force push the destination branch to the source branch...

Sigh, guess what I just did...
